src/main/kotlin/minesweeper/domain/Board.kt

```diff
@@ -5,87 +5,119 @@ import minesweeper.domain.block.Block
 import minesweeper.domain.block.Mine
 import minesweeper.domain.block.None
 import minesweeper.domain.block.Position
+import minesweeper.domain.block.replace
+import minesweeper.domain.game.GameResult
+import minesweeper.domain.game.State
 
 data class Board(val area: Area, val blocks: List<Block>) {
     init {
-        require(blocks.count { it is Mine } > MINIMUM_MINE_COUNT)
-        require(area.getArea() == blocks.size)
+        require(blocks.count { it is Mine } > MINIMUM_MINE_COUNT) { MINIMUM_MINE_REQUIRED }
+        require(area.getArea() == blocks.size) { AREA_BLOCK_SIZE_SHOULD_SAME }
     }
 
     fun scanMines(): Board {
-        var blocksResult = emptyList<Block>()
+        val blocksResult = mutableListOf<Block>()
         for (block in blocks) {
-            var mineCount = 0
-            mineCount = findMineCount(mineCount, block)
-            blocksResult = addNoneBlock(block, blocksResult, mineCount)
-            blocksResult = addMineBlock(block, blocksResult)
+            addBlocks(block, blocksResult)
         }
         return Board(area, blocksResult)
     }
 
-    private fun addMineBlock(
-        block: Block,
-        blockResult: List<Block>
-    ): List<Block> {
-        var blocks = blockResult
-        if (block is Mine) {
-            blocks = blocks + block
+    fun scanMine(x: Int, y: Int): GameResult {
+        val targetBlock = findBlock(x, y) ?: return GameResult(State.PLAY, this)
+        if (targetBlock.isMine()) return GameResult(State.LOSE, this)
+        var blockResult = blocks
+        for (position in MOVABLE_POSITION) {
+            val newX = x + position.x
+            val newY = y + position.y
+            blockResult = updateBlocks(blockResult, newX, newY)
         }
-        return blocks
+        val mineCount = blocks.count { it is Mine }
+        val hasVisitedCount = blockResult.count() { it.hasVisited() }
+        val remainCount = blocks.size - hasVisitedCount
+        if (mineCount == remainCount) {
+            return GameResult(State.WIN, Board(area, blockResult))
+        }
+        return GameResult(State.PLAY, Board(area, blockResult))
     }
 
     fun findBlock(x: Int, y: Int): Block? {
         val targetPosition = Position(x, y)
         return blocks.find { it.getPosition() == targetPosition }
     }
 
-    private fun findIsMine(block: Block?): Boolean {
-        return block is Mine
+    private fun updateBlocks(
+        blockResult: List<Block>,
+        newX: Int,
+        newY: Int
+    ): List<Block> {
+        var blockResult1 = blockResult
+        val nearBlock = findBlock(newX, newY)
+        if (nearBlock != null && nearBlock is None) {
+            blockResult1 =
+                blockResult1.replace(nearBlock.updateBlock(findNearMineCount(newX, newY))) { it == nearBlock }
+        }
+        return blockResult1
```

Comment:
```suggestion
        val nearBlock = findBlock(newX, newY)
        if (nearBlock != null && nearBlock is None) {
            return blockResult.replace(nearBlock.updateBlock(findNearMineCount(newX, newY))) { it == nearBlock }
        }
        return blockResult
```
변수의 재할당을 하지 않아도 이런식으로 로직을 구성해볼 수 있을 것 같습니다.
지역변수에 재할당을 계속하면, 해당 변수가 어떻게 재할당이 되는지 추적해야하는 공수비용이 들어 로직을 파악하기 어렵다고 생각이 됩니다.
최대한 재할당을 피해서 로직을 구성해보는 것은 어떨까요?